lib/SILOptimizer/FunctionSignatureTransforms/ExistentialTransform.cpp

```diff
@@ -0,0 +1,590 @@
+  auto FTy = F->getLoweredFunctionType();
+  SILModule &M = F->getModule();
+  auto &Ctx = M.getASTContext();
+  GenericSignature *NewGenericSig;
+  GenericEnvironment *NewGenericEnv;
+
+  // Form a new generic signature based on the old one.
+  GenericSignatureBuilder Builder(Ctx);
+
+  /// If the original function is generic, then maintain the same.
+  auto OrigGenericSig = FTy->getGenericSignature();
+  /// First, add the old generic signature.
+  Builder.addGenericSignature(OrigGenericSig);
+
+  /// Convert existential argument types to generic argument types.
+  convertExistentialArgTypesToGenericArgTypes(Builder);
+
+  /// Compute the updated generic signature.
+  NewGenericSig = std::move(Builder).computeGenericSignature(
+      SourceLoc(), /*allowConcreteGenericParams=*/true);
+  NewGenericEnv = NewGenericSig->createGenericEnvironment();
+
+  /// Create a lambda for GenericParams.
+  auto getCanonicalType = [&](Type t) -> CanType {
+    return t->getCanonicalType(NewGenericSig);
+  };
+
+  /// Original list of parameters
+  SmallVector<SILParameterInfo, 4> params;
+  params.append(FTy->getParameters().begin(), FTy->getParameters().end());
+
+  /// Create the complete list of parameters.
+  int Idx = 0;
+  llvm::SmallVector<SILParameterInfo, 8> InterfaceParams;
+  InterfaceParams.reserve(params.size());
+  for (auto &param : params) {
+    auto iter = ArgToGenericTypeMap.find(Idx);
+    if (iter != ArgToGenericTypeMap.end()) {
+      auto GenericParam = iter->second;
+      InterfaceParams.push_back(SILParameterInfo(getCanonicalType(GenericParam),
+                                                 param.getConvention()));
+    } else {
+      auto Ty = param.getType();
+      InterfaceParams.push_back(
+          SILParameterInfo(!Ty->hasTypeParameter()
+                               ? Ty->mapTypeOutOfContext()->getCanonicalType()
+                               : Ty,
+                           param.getConvention()));
+    }
+    Idx++;
+  }
+
+  /// Now add the result types.
+  llvm::SmallVector<SILResultInfo, 8> InterfaceResults;
+  for (auto &result : FTy->getResults()) {
+    auto Ty = result.getType();
+    InterfaceResults.push_back(result.getWithType(
+        !Ty->hasTypeParameter() ? Ty->mapTypeOutOfContext()->getCanonicalType()
+                                : Ty));
+  }
+
+  /// Now add the errors.
+  Optional<SILResultInfo> InterfaceErrorResult;
+  if (FTy->hasErrorResult()) {
+    auto errorResult = FTy->getErrorResult();
+    auto Ty = errorResult.getType();
+    InterfaceErrorResult = SILResultInfo(
+        !Ty->hasTypeParameter() ? Ty->mapTypeOutOfContext()->getCanonicalType()
+                                : Ty,
+        errorResult.getConvention());
+  }
+
+  /// Now add the yields.
+  llvm::SmallVector<SILYieldInfo, 8> InterfaceYields;
+  for (SILYieldInfo yield : FTy->getYields()) {
+    auto Ty = yield.getType();
+    InterfaceYields.push_back(yield.getWithType(
```

Comment:
You can replace this whole thing with InterfaceYields.push_back(yield). You don't need to do anything with `Ty`.